amy/fiscal/tests/test_filters.py

```diff
@@ -0,0 +1,454 @@
+        cls.organization = Organization.objects.create(
+            fullname="Test Organization", domain="example.org"
+        )
+        cls.membership = Membership.objects.create(
+            name="Test Membership",
+            variant="Bronze",
+            agreement_start=date.today(),
+            agreement_end=date.today(),
+            contribution_type="Financial",
+            public_instructor_training_seats=1,
+        )
+        cls.member = Member.objects.create(
+            membership=cls.membership, organization=cls.organization, role=member_role
+        )
+        # create a test membership that sphould be filtered out
+        cls.organization2 = Organization.objects.create(
+            fullname="To Be Filtered Out", domain="example2.org"
+        )
+        cls.membership2 = Membership.objects.create(
+            name="To Be Filtered Out",
+            variant="Silver",
+            agreement_start=date(2015, 1, 1),
+            agreement_end=date(2016, 1, 1),
+            contribution_type="Person-days",
+            public_instructor_training_seats=0,
+        )
+        cls.member2 = Member.objects.create(
+            membership=cls.membership2, organization=cls.organization2, role=member_role
+        )
+        cls.qs = Membership.objects.all().annotate(
+            instructor_training_seats_total=(
+                # Public
+                F("public_instructor_training_seats")
+                + F("additional_public_instructor_training_seats")
+                # Coalesce returns first non-NULL value
+                + Coalesce("public_instructor_training_seats_rolled_from_previous", 0)
+                # Inhouse
+                + F("inhouse_instructor_training_seats")
+                + F("additional_inhouse_instructor_training_seats")
+                + Coalesce("inhouse_instructor_training_seats_rolled_from_previous", 0)
+            ),
+            instructor_training_seats_remaining=(
+                # Public
+                F("public_instructor_training_seats")
+                + F("additional_public_instructor_training_seats")
+                # Coalesce returns first non-NULL value
+                + Coalesce("public_instructor_training_seats_rolled_from_previous", 0)
+                - Count(
+                    "task", filter=Q(task__role__name="learner", task__seat_public=True)
+                )
+                - Coalesce("public_instructor_training_seats_rolled_over", 0)
+                # Inhouse
+                + F("inhouse_instructor_training_seats")
+                + F("additional_inhouse_instructor_training_seats")
+                + Coalesce("inhouse_instructor_training_seats_rolled_from_previous", 0)
+                - Count(
+                    "task",
+                    filter=Q(task__role__name="learner", task__seat_public=False),
+                )
+                - Coalesce("inhouse_instructor_training_seats_rolled_over", 0)
+            ),
+        )
```

Comment:
Since it looks like we have this same query in multiple places (twice in this file, one in `fiscal.views.AllMemberships.queryset` and one in `reports.views.membership_trainings_stats`) then I propose to extract it out, for example into a utility function, and use it in all 4 places instead. Hopefully this would lead to less changes in future and less errors in the calculations since they are quite difficult.

Reply:
Good idea. I moved this to a new Manager for the Membership object - could do with another review